signer/kms/aws/signer_test.go

```diff
@@ -0,0 +1,374 @@
+			hash:    crypto.SHA512,
+			ref:     "awskms:///arn:aws-us-gov:kms:us-gov-west-1:111122223333:key/1234abcd-12ab-34cd-56ef-1234567890ab",
+			mess:    []string{"foo", "bar", "baz"},
+			wantErr: false,
+		},
+		{
+			name:        "bad ref",
+			hash:        crypto.SHA256,
+			ref:         "blablabla",
+			mess:        []string{"foo", "bar", "baz"},
+			wantErr:     true,
+			expectedErr: fmt.Errorf("kms specification should be in the format awskms://[ENDPOINT]/[ID/ALIAS/ARN] (endpoint optional)"),
+		},
+		{
+			name:        "unsupported hash algorithm",
+			hash:        crypto.RIPEMD160,
+			ref:         "awskms:///arn:aws-us-gov:kms:us-gov-west-1:111122223333:key/1234abcd-12ab-34cd-56ef-1234567890ab",
+			mess:        []string{"foo", "bar", "baz"},
+			wantErr:     true,
+			expectedErr: fmt.Errorf(`unsupported hash algorithm: "RIPEMD-160" not in [SHA-256 SHA-384 SHA-512]`),
+		},
+		{
+			name:        "gcp ref",
+			hash:        crypto.SHA256,
+			ref:         "gcpkms://projects/testproject-23231/locations/europe-west2/keyRings/test/cryptoKeys/test",
+			mess:        []string{"foo", "bar", "baz"},
+			wantErr:     true,
+			expectedErr: fmt.Errorf("kms specification should be in the format awskms://[ENDPOINT]/[ID/ALIAS/ARN] (endpoint optional)"),
+		},
+	}
+
+	for _, tt := range tests {
+		errFound := false
+		fmt.Println("verify test: ", tt.name)
+		ctx := context.TODO()
+		ksp := kms.New(kms.WithRef(tt.ref), kms.WithHash(tt.hash.String()))
+		c, err := newFakeAWSClient(context.TODO(), ksp)
+		if tt.wantErr && err != nil {
+			errFound = true
+			assert.ErrorAs(t, err, &tt.expectedErr)
+			continue
+		} else if err != nil {
+			t.Fatal(err)
+		}
+
+		for _, mess := range tt.mess {
+			bs, bv, err := createTestKey()
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			dig, _, err := cryptoutil.ComputeDigest(bytes.NewReader([]byte(mess)), tt.hash, awsSupportedHashFuncs)
+			if tt.wantErr && err != nil {
+				errFound = true
+				assert.ErrorAs(t, err, &tt.expectedErr)
+				continue
+			} else if err != nil {
+				t.Fatal(err)
+			}
+
+			sig, err := c.sign(ctx, []byte(dig), crypto.SHA256)
+			if tt.wantErr && err != nil {
+				errFound = true
+				assert.ErrorAs(t, err, &tt.expectedErr)
+				continue
+			} else if err != nil {
+				t.Fatal(err)
+			}
+
+			bsig, err := bs.Sign(bytes.NewReader([]byte(dig)))
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			err = c.verifyRemotely(ctx, sig, dig)
+			if tt.wantErr && err != nil {
+				errFound = true
+				assert.ErrorAs(t, err, &tt.expectedErr)
+				continue
+			} else if err != nil {
+				t.Fatal(err)
+			}
+
+			err = c.verifyRemotely(ctx, bsig, dig)
+			if err == nil {
+				t.Fatal("expected verification to fail")
+			}
+
+			err = bv.Verify(bytes.NewReader([]byte(dig)), sig)
+			if err == nil {
+				t.Fatal("expected verification to fail")
+			}
+
+		}
+
+		if tt.wantErr && !errFound {
+			t.Fatalf("expected test %s to fail with error %s", tt.name, tt.expectedErr.Error())
+		}
+	}
+}
```

Comment:
I really love the test patterns you used here

Reply:
huh - I had imposter syndrome that this was terrible! This comment made me feel good <3